### PR TITLE
feat: Add breakpoints tokens and script to generate SCSS media conditions

### PR DESCRIPTION
**Pull request by aaronccasanova**
### WHAT is this pull request doing?

This PR introduces a new `breakpoints.json` token group and a script to generate a stylesheet containing aliased media conditions.

https://user-images.githubusercontent.com/32409546/164096731-f692bf0b-8a72-43f3-b2c3-41cc892b71e4.mov

> The box below the media condition is `green` when `true` and `orange` when `false`

**Storybook demo of [all aliased media conditions](https://5d559397bae39100201eedc1-tlzloilcdq.chromatic.com/iframe.html?id=playground-playground--playground&args=&viewMode=story)**

### Steps

- Added a `breakpoints.json` token group with `xs`, `sm`, `md`, `lg`, and `xl` token aliases.
- Added the boilerplate setup to generate `--p-breakpoints-*` custom properties
- Created a `polaris-react/build/gen-media-queries.js` script to generate a `src/styles/_media-queries.scss` stylesheet containing SCSS variables for aliased media conditions e.g. `$p-breakpoints-md-up`

<details>
  <summary>Example of the generated <code>_media-queries.scss</code> file:</summary>
  
```scss
/* stylelint-disable unit-disallowed-list */
/* stylelint-disable length-zero-no-unit */

/* Breakpoints - Aliases */
$p-breakpoints-xs: 0em;
$p-breakpoints-sm: 35em;
$p-breakpoints-md: 48em;
$p-breakpoints-lg: 63em;
$p-breakpoints-xl: 111em;

/* Breakpoints - Media conditions */
$p-breakpoints-xs-up: '(min-width: 0em)';
$p-breakpoints-xs-down: '(max-width: 0em)';
$p-breakpoints-xs-only: '(min-width: 0em) and (max-width: 35em)';

$p-breakpoints-sm-up: '(min-width: 35em)';
$p-breakpoints-sm-down: '(max-width: 35em)';
$p-breakpoints-sm-only: '(min-width: 35em) and (max-width: 48em)';

$p-breakpoints-md-up: '(min-width: 48em)';
$p-breakpoints-md-down: '(max-width: 48em)';
$p-breakpoints-md-only: '(min-width: 48em) and (max-width: 63em)';

$p-breakpoints-lg-up: '(min-width: 63em)';
$p-breakpoints-lg-down: '(max-width: 63em)';
$p-breakpoints-lg-only: '(min-width: 63em) and (max-width: 111em)';

$p-breakpoints-xl-up: '(min-width: 111em)';
$p-breakpoints-xl-down: '(max-width: 111em)';
$p-breakpoints-xl-only: '(min-width: 111em)';
```
</details>

### Design

The design of the generated media conditions is as close as I could get to the up coming [`@custom-media` spec](https://www.w3.org/TR/mediaqueries-5/#at-ruledef-custom-media) without introducing the [`postcss-custom-media` plugin](https://github.com/postcss/postcss-custom-media). The reason being we do not want to introduce a dependency on a `postcss` plugin to use Polaris.

**Example usage:**
```scss
@import '../styles/media-queries'
// @import '@shopify/polaris/styles/media-queries' // Future public API

.container {
  padding: 20px;

  @media #{$p-breakpoints-md-up} {
    padding: 40px;
  }
}
```

The hope is that this will create the most seamless upgrade path in the future when we remove these aliased media conditions in favor of CSS custom media variables e.g.

**Before:**
```scss
@media #{$p-breakpoints-md-up} {/*...*/}
```

**After:**
```scss
@media (--p-breakpoints-md-up) {/*...*/}
```

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
- [ ] For visual design changes, ping @ sarahill to update the Polaris UI kit

### CODE REVIEW

File: polaris-react/src/styles/_tokens.scss [hunk 1]
```diff
@@ -0,0 +1,32 @@
+/* stylelint-disable unit-disallowed-list */
```
Comment (aaronccasanova) [inline polaris-react/src/styles/_tokens.scss]: TODO:
- Do NOT commit this file - Add to `.gitignore`
- Instead generate it in a `prebuild` script

File: polaris-react/playground/Playground.scss [hunk 2]
```diff
@@ -0,0 +1,103 @@
+@import '../src/styles/tokens';
```
Comment (aaronccasanova) [inline polaris-react/playground/Playground.scss]: Note to self.. Do NOT commit Playground changes.

Committed to have access to the hosted storybook playground and give reviewers some reference examples.

File: polaris-react/playground/Playground.tsx [hunk 3]
```diff
@@ -1,11 +1,48 @@
 import React from 'react';
 
 import {Page} from '../src';
+import breakpoints from '../src/tokens/token-groups/breakpoints.json';
```
Comment (aaronccasanova) [inline polaris-react/playground/Playground.tsx]: Same as above.. https://github.com/Shopify/polaris/pull/5558/files#r853500178

File: polaris-react/src/styles/_tokens.scss [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+$p-breakpoints-xs-em: 0em;
+
+$p-breakpoints-sm: 480px;
+$p-breakpoints-sm-em: 30em;
+
+$p-breakpoints-md: 768px;
+$p-breakpoints-md-em: 48em;
+
+$p-breakpoints-lg: 1008px;
+$p-breakpoints-lg-em: 63em;
```
Comment (aaronccasanova) [inline polaris-react/src/styles/_tokens.scss]: You might be wondering why I have included both `px` and `em` breakpoint aliases.. My thinking was that users may need to compose breakpoints (such as: `$between-xs-and-md` or `$mobile-nav-collapsed`) and these variables would serve as the low level building blocks. Thoughts?

Comment (alex-page) [reply polaris-react/src/styles/_tokens.scss]: I would prefer providing the best and simplest solution. I don't think we should provide the `px` breakpoints.

Comment (aaronccasanova) [reply polaris-react/src/styles/_tokens.scss]: Sounds good, I'll remove the `px` aliases. We really only need the `em` aliases for the above note about composition.

Comment (aaronccasanova) [reply polaris-react/src/styles/_tokens.scss]: Updated: https://github.com/Shopify/polaris/pull/5558/commits/e6642873d983fd6694bec4091d3e2be816c2cfc3

<details>
  <summary>Example of generated output</summary>
  
 ```scss
 /* stylelint-disable unit-disallowed-list */
/* stylelint-disable length-zero-no-unit */

/* Breakpoints - Aliases */
$p-breakpoints-xs: 0em;
$p-breakpoints-sm: 30em;
$p-breakpoints-md: 48em;
$p-breakpoints-lg: 63em;

/* Breakpoints - Media conditions */
$p-breakpoints-xs-up: '(min-width: 0em)';
$p-breakpoints-xs-down: '(max-width: 0em)';
$p-breakpoints-xs-only: '(min-width: 0em) and (max-width: 30em)';

$p-breakpoints-sm-up: '(min-width: 30em)';
$p-breakpoints-sm-down: '(max-width: 30em)';
$p-breakpoints-sm-only: '(min-width: 30em) and (max-width: 48em)';

$p-breakpoints-md-up: '(min-width: 48em)';
$p-breakpoints-md-down: '(max-width: 48em)';
$p-breakpoints-md-only: '(min-width: 48em) and (max-width: 63em)';

$p-breakpoints-lg-up: '(min-width: 63em)';
$p-breakpoints-lg-down: '(max-width: 63em)';
$p-breakpoints-lg-only: '(min-width: 63em)';
 ```
</details>

File: polaris-react/package.json [hunk 5]
```diff
@@ -49,6 +49,8 @@
     "test": "loom test",
     "check": "npm-run-all lint type-check test",
     "clean": "rimraf build build-internal",
+    "gen-tokens-scss": "node ./scripts/build-tokens-scss.js",
```
Comment (aaronccasanova) [inline polaris-react/package.json]: Note: This and all the pre* scripts would eventually be removed in favor of importing the file directly from `@shopify/polaris-tokens`. The idea being, `@shopify/polaris-tokens` would include this as part of it's build process and consumers (e.g. `polaris-react`, `web`, etc.) would consume the breakpoint media conditions directly in their SCSS stylesheets e.g.
```scss
@import '@shopify/polaris-tokens/styles/media';

.my-app {
  padding: 20px;
  
  @media #{$p-breakpoints-md-up} {
    padding: 40px;
  }
}
```

File: polaris-react/playground/Playground.scss (hunk 2 above)
Comment (kyledurand) [reply polaris-react/playground/Playground.scss]: To tell git to ignore the playground:
`git update-index --skip-worktree polaris-react/playground/Playground.tsx`

To tell it to not ignore the playground:
`git update-index --no-skip-worktree polaris-react/playground/Playground.tsx`

File: polaris-react/src/styles/_tokens.scss (hunk 1 above)
Comment (aaronccasanova) [reply polaris-react/src/styles/_tokens.scss]: Done

File: polaris-react/playground/Playground.scss (hunk 2 above)
Comment (aaronccasanova) [reply polaris-react/playground/Playground.scss]: Reverted: https://github.com/Shopify/polaris/pull/5558/commits/80ccf6c7f6516d511681f08d6d9eb608ce98853e

File: polaris-react/playground/Playground.tsx (hunk 3 above)
Comment (aaronccasanova) [reply polaris-react/playground/Playground.tsx]: Reverted: https://github.com/Shopify/polaris/pull/5558/commits/80ccf6c7f6516d511681f08d6d9eb608ce98853e

File: polaris-react/package.json (hunk 5 above)
Comment (aaronccasanova) [reply polaris-react/package.json]: After syncing with @alex-page, we decided to keep the SCSS media query generation in `@shopify/polaris` and the usage internal (as we don't need to focus on the public API at this phase). Additionally, the `breakpoints` tokens are okay to be moved over to `@shopify/polaris-tokens` and transformed to `--p-breakpoints-*` custom properties.

**Review by lgriffee — APPROVED** [review]
💯🚢✨🎉

**Comment by lgriffee** [comment]
Do we want to merge this into `main` or our `v10.0.0-major` branch? I guess this isn't technically a breaking change since we're just adding new values and not removing any old ones...

**Review by alex-page — APPROVED** [review]

File: polaris-tokens/src/token-groups/breakpoints.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
+export const breakpoints = {
+  'breakpoints-xs': {
+    value: '0px',
+    description: '',
+  },
```
Comment (alex-page) [inline polaris-tokens/src/token-groups/breakpoints.ts:5]: Random question but is `0` a needed breakpoint?

Comment (aaronccasanova) [reply polaris-tokens/src/token-groups/breakpoints.ts:5]: Yes, we follow a mobile first approach so while this value is `0px`, `xs` means `0px-559px`:
```
| 0px ->
| xs
         | 560px ->
         | sm
                    | 768px ->
                    | md
                               | 1008px ->
                               | lg
                                           | 1776px ->
                                           | xl
```

This consistent with every design system I researched:
```
Primer:
  xs: 0px
  sm: 544px
  md: 768px
  lg: 1012px
  xl: 1280px
MUI:
  xs: 0px
  sm: 600px
  md: 900px
  lg: 1200px
  xl: 1536px
Bootstrap:
  xs: 0px
  sm: 576px
  md: 768px
  lg: 992px
  xl: 1200px
  xxl: 1400px
  ```

**Comment by aaronccasanova** [comment]
> Do we want to merge this into main or our v10.0.0-major branch? I guess this isn't technically a breaking change since we're just adding new values and not removing any old ones...

Exactly @lgriffee! This is a feature enhancement that is currently only available internally. Thus it is safe to merge directly into `main`.

File: polaris-tokens/src/token-groups/breakpoints.ts (hunk 6 above)
Comment (alex-page) [reply polaris-tokens/src/token-groups/breakpoints.ts:5]: Amazing! I don't know why I just wanted to check this lol. Thanks @aaronccasanova 💯

**Pull request merged**